Added gps connection

AsafL910 opened the pull request:
Added gps connection with handling of permissions - also fixed pr comments from last time and refactored a lot of logic
![WhatsApp Image 2025-02-06 at 17 02 07 (1)](https://github.com/user-attachments/assets/855ce6a7-cb79-45e6-9343-633162a64c42)
![WhatsApp Image 2025-02-06 at 17 02 07](https://github.com/user-attachments/assets/8a4e3c61-f010-40b6-a374-3f79e9643f3b)
![WhatsApp Image 2025-02-06 at 17 02 08](https://github.com/user-attachments/assets/fd272495-c031-4198-b3f7-079195d2527f)
![WhatsApp Image 2025-02-06 at 17 03 24](https://github.com/user-attachments/assets/6994a052-89d3-41b2-99ce-3a417d1ad7b7)

AsafL910 committed:
Added polygon to the map when fetching data

still does not work with parsing multiPolygon areas from the api like Italy which has islands

AsafL910 committed:
Add the ability to parse multipolygons and show them on the map

AsafL910 committed:
Fixed source creation to be on map start

AsafL910 committed:
split map setup to different functions

as well as used color theme in layer color

AsafL910 committed:
Added "Add Location" button

- it will be functional later when the user will be handled and saved
- also added basic interaction with map click on location event

AsafL910 committed:
fixed search result design

AsafL910 committed:
pr fixes

AsafL910 committed:
Exported logic into utils files

AsafL910 committed:
Moved more logic into MapUtils

AsafL910 committed:
Added Greyscale effect to map like in design

AsafL910 committed:
added basic gps connection

AsafL910 committed:
WIP on added-gps-connection

 what's left is to
:-manage edge cases with gps connections
- implement toast component in every new fragment created

AsafL910 committed:
Fixed some location bugs and added back loading circle

AsafL910 committed:
Fixed map to be higher resolution

AsafL910 committed:
Fixed location not showing when first allowing permissions

AsafL910 committed:
hide keyboard when search result is clicked

also added some pr fixes from before

AsafL910 committed:
Fixed search component

- it now works with place_id which is more accurate
- the design is improved and looks more like the initial figma design

AsafL910 committed:
fixed zoom too strong when fetching user location from api

AsafL910 committed:
Fixed icons to be linked to the theme color

AsafL910 committed:
pr fixes

RonyL02 commented on app/src/main/java/com/col/eventradar/api/dto/LocationDetailsResultDTO.kt:
```diff
@@ -0,0 +1,130 @@
+                }
+
+                else -> {
+                    // Fallback to centroid point
```
remove comment

RonyL02 commented on app/src/main/java/com/col/eventradar/api/dto/LocationDetailsResultDTO.kt:
```diff
@@ -0,0 +1,130 @@
+    val osm_id: Long,
+    val category: String,
+    val type: String,
+    val admin_level: Int,
```
transform all the snake case into pascal case

LiorVainer replied:
to camelCase I think is better suited for data classes

RonyL02 replied:
yeah thats what i meant

RonyL02 commented on app/src/main/java/com/col/eventradar/api/dto/LocationSearchResultDTO.kt:
```diff
@@ -0,0 +1,37 @@
+    @SerializedName("country") val country: String?
+)
+
+fun LocationSearchResultDTO.toDomain(): LocationSearchResult {
```
```suggestion
fun LocationSearchResultDTO.toModel(): LocationSearchResult {
```

RonyL02 commented on app/src/main/java/com/col/eventradar/api/dto/LocationDetailsResultDTO.kt:
```diff
@@ -0,0 +1,130 @@
+    val geometry: Geometry,
+    val icon: String
+) {
+        fun toMapLibreFeature(): Feature {
```
i think you should move it to an extention function to keep the data class clean, as well as to make some separation between the dto and maplibre. maybe put it under map utils for example

RonyL02 commented on app/src/main/java/com/col/eventradar/ui/LocationSearchFragment.kt:
```diff
@@ -0,0 +1,159 @@
+        gpsFragment = GpsLocationSearchFragment()
+
+        childFragmentManager.beginTransaction()
+            .add(R.id.gpsContainer, gpsFragment!!, "GpsFragmentTag")
```
```suggestion
            .add(R.id.gpsContainer, gpsFragment!!, GpsLocationSearchFragment.TAG)
```

RonyL02 commented on app/src/main/java/com/col/eventradar/ui/components/ToastFragment.kt:
```diff
@@ -0,0 +1,13 @@
+import android.widget.Toast
+
+class ToastFragment : Fragment() {
+    fun showToast(message: String, duration: Int = Toast.LENGTH_SHORT) {
```
you still use Toast.makeText instead of using the fragment

RonyL02 replied:
if its the only method, you can override the invoke operator instead of explicitly calling showToast every time

  operator fun invoke(message: String, duration: Int = Toast.LENGTH_SHORT) {
        ...
    } 

and instead of toastFragment.makeToast(...)
all you gotta do is toastFragment(...)

RonyL02 commented on app/src/main/java/com/col/eventradar/ui/views/MapFragment.kt:
```diff
@@ -1,103 +1,131 @@
         bindingInternal = FragmentMapBinding.inflate(inflater, container, false)
         binding.mapView.onCreate(savedInstanceState)
+        locationFragment = GpsLocationMapFragment()
+        childFragmentManager.beginTransaction().add(locationFragment, "GpsLocationMapFragment")
```
```suggestion
        childFragmentManager.beginTransaction().add(locationFragment, GpsLocationMapFragment.TAG)
```

RonyL02 commented on app/src/main/java/com/col/eventradar/ui/views/MapFragment.kt:
```diff
@@ -1,103 +1,131 @@
+            .commit()
+
+        toastFragment = ToastFragment()
+        childFragmentManager.beginTransaction().add(toastFragment, "ToastFragment").commit()
```
```suggestion
        childFragmentManager.beginTransaction().add(toastFragment, ToastFragment.TAG).commit()
```

RonyL02 commented on app/src/main/java/com/col/eventradar/ui/views/MapFragment.kt:
```diff
@@ -1,103 +1,131 @@
-                .zoom(zoom ?: DEFAULT_ZOOM)
-                .build()
+            this.map = map
+            map.setStyle(MapUtils.DEFAULT_MAP_STYLE_URL) { style ->
```
after calling setStyle isnt the style available in the map via map.style? why pass both map and style separately in attachToMap instead of just passing the map?

AsafL910 replied:
In MapLibre you need to access the map style inside a callback function because loading the style is asynchronous. When you set a style with map.setStyle() the map needs time to download and apply the style before it is ready for use.

RonyL02 commented on app/src/main/java/com/col/eventradar/ui/views/MapFragment.kt:
```diff
@@ -1,103 +1,131 @@
+        locationFragment.locationComponent?.lastKnownLocation?.let {
+            val cameraPosition = CameraPosition.Builder()
+                .target(LatLng(it.latitude, it.longitude))
+//                .zoom(10.0)
```
remove

RonyL02 commented on app/src/main/java/com/col/eventradar/utils/MapUtils.kt:
```diff
@@ -0,0 +1,132 @@
+
+    const val TileJSON_VERSION = "2.1.0"
+
+    val TAG = "MapUtils"
```
is it necessary?

AsafL910 replied:
it is used in logs further down this file

LiorVainer commented on app/build.gradle.kts:
```diff
@@ -40,6 +40,9 @@ android {
 }
 
 dependencies {
+    implementation("com.google.code.gson:gson:2.10")
```
Cant it be in version catalog like rest libs?

LiorVainer commented on app/build.gradle.kts:
```diff
@@ -40,6 +40,9 @@ android {
 
 dependencies {
+    implementation("com.google.code.gson:gson:2.10")
+    implementation("com.squareup.retrofit2:retrofit:2.9.0")
```
Same about all libs with string values in implementation function

LiorVainer replied:
usually android studio suggest to switch to version catalog. if not Tal gave example in class to how to make it manually

LiorVainer commented on app/src/main/java/com/col/eventradar/ui/LocationSearchFragment.kt:
```diff
@@ -0,0 +1,159 @@
+            ?: throw RuntimeException("$parentFragment must implement MapFragmentListener")
+    }
+
+    private fun searchLocation(query: String) {
```
Dont make this function right in the fragment it self but rather use `ViewModel` extended custom class for it.
See the "Movie Reviewer" in the **Moodle** of the course in "example apps" section. the code in this app is really explains it self and if it in the "example apps" section it is probably a good example to how make this the right way

![image](https://github.com/user-attachments/assets/75d79607-bfdc-4e8b-aed1-2ace6f203919)

AsafL910 committed:
pr fixes

- mainly moved lib versions to version catalog

AsafL910 committed:
more pr fixes

changed to MVVM and fixed more comments

AsafL910 committed:
changed all toast references to fragment

AsafL910 committed:
removed some unused data from dto

also fixed names to be camelCase as Gson can automatically map snace_case to camelCase when parsing

AsafL910 committed:
Merge remote-tracking branch 'origin/main' into added-gps-connection

AsafL910 committed:
fixed merge conflict

LiorVainer commented on app/src/main/java/com/col/eventradar/ui/viewmodels/LocationSearchViewModel.kt:
```diff
@@ -0,0 +1,38 @@
+            hasSearched = true
+            _isLoading.value = true
+            try {
+                val results = OpenStreetMapService.api.searchLocation(query = query, limit = 5)
```
Final comment in this topic: I would extract to `repository` class this api call so there will be a layer bewteen ViewModel and directly calling api. saw this also on the "example apps".  I made example for that my self.

```kotlin
class EventRepository {
    suspend fun fetchEvents(
        fromDate: LocalDateTime?,
        toDate: LocalDateTime?,
        alertLevels: List<AlertLevel>?,
        eventTypes: List<EventType>?,
        country: String?,
    ): EventListResponseDTO? {
        val queryParams =
            GdacsQueryHelper.buildEventQueryParams(
                fromDate = fromDate ?: LocalDateTime.now().minusDays(7),
                toDate = toDate ?: LocalDateTime.now(),
                alertLevels = alertLevels ?: listOf(AlertLevel.RED, AlertLevel.ORANGE),
                eventTypes = eventTypes ?: EventType.allExceptUnknown,
                country = country ?: "United States",
            )

        Log.d("EventRepository", "Query params: $queryParams")

        return try {
            val response =
                GdacsService.api.getEventList(
                    fromDate = queryParams[GdacsQueryParams.FROM_DATE] ?: return null,
                    toDate = queryParams[GdacsQueryParams.TO_DATE] ?: return null,
                    alertLevel = queryParams[GdacsQueryParams.ALERT_LEVEL] ?: return null,
                    eventList = queryParams[GdacsQueryParams.EVENT_LIST] ?: return null,
                    country = queryParams[GdacsQueryParams.COUNTRY] ?: return null,
                )

            if (response.isSuccessful) {
                val body = response.body()
                if (body != null && body.features.isNotEmpty()) {
                    body
                } else {
                    Log.w("EventRepository", "No events found" + response.message())
                    null
                }
            } else {
                Log.e("EventRepository", "API Error: ${response.code()} - ${response.message()}")
                null
            }
        } catch (e: Exception) {
            Log.e("EventRepository", "Network error fetching events", e)
            null
        }
    }
}
```

```kotlin
interface GdacsApi {
    @GET("geteventlist/SEARCH")
    suspend fun getEventList(
        @Query("fromDate") fromDate: String,
        @Query("toDate") toDate: String,
        @Query("alertlevel") alertLevel: String = "orange;red",
        @Query("eventlist") eventList: String = "EQ;TS,TC,FL,VO,DR,WF",
        @Query("country") country: String = "United States",
    ): Response<EventListResponseDTO>
}
```

also wrapping response of `ApiService` in `Retrofit` response class is pretty useful for handling status codes and empty response and such things. is similar to `AxiosResponse`

RonyL02 replied:
does it make sense to call  it a repository though? 
i think EventUseCase/GetEventUseCase might be more suitable
you can look it up its a pretty common practice

LiorVainer replied:
class named `GetEventUseCase ` ?

LiorVainer replied:
Leave it right now and Merge MR. we will decide that later
